### ISSUE TITLE
Factorial

### ISSUE THREAD

**Issue by Merlin1846** [godotengine/godot-proposals#5882]
### Describe the project you are working on

Right now an n-dimensional wave function collapse implementation but this is a more general purpose.

### Describe the problem or limitation you are having in your project

It's a basic math function, it should be included in the engine. I would add this myself but the closest language to c++ I've used is Rust and when I tried to add the code I messed up with the godot/core/math/math_funcs.h file which I believe is where this would go.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

You won't have to do it you're self or make a global math class because Godot doesn't have factorial.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

```
func factorial(value: int) -> int:
	var out: int = 1
	for i in abs(value):
		out *= i + 1
	return out
```

That simple

### If this enhancement will not be used often, can it be worked around with a few lines of script?

This is a math function some people will use others won't but considering how many are already implemented I believe it makes sense to add this.

### Is there a reason why this should be core and not an add-on in the asset library?

It's just a minor thing like sin or round.

**Comment by Merlin1846** [comment]
Yes I know I should learn c++ but I already procrastinate too much and don't need another excuse to not work on my projects.

**Comment by AThousandShips** [comment]
Should be relatively straight forward, can take a look at this, though the properly useful case would be for floating point as `int64_t` only holds 21!

**Comment by AThousandShips** [comment]
Given that this is likely not going to be used by most users, and that unlike for example `sin`/`cos`/etc. there's no implementation already provided in the standard library, I don't think there's much support for adding this, considering how easy it is to implement on your own.

Adding it beyond the most trivial implementation would add quite a bit of data to the binaries, when compared to these considerations.

**Comment by AThousandShips** [comment]
The only real reason I can think of to add it to core would be performance, as implementing it in GDscript would be lacking compared to even the most naive implementation in C++, and for that I would suggest simply creating a lookup table, given the tiny number of values you can fit in any of the data types (20 in `int`, 170 in `float`).

There would also be the consideration of accuracy, for example the ability to compute `double` factorial values in `long double`, which again is not possible with GDscript, but that's such a specific consideration I feel it's not reason enough to put it in core.
